iroh-gateway/src/headers.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{constants::*, response::ResponseFormat};
+use axum::http::header::*;
+use chrono::prelude::Utc;
+use cid::Cid;
+
+#[tracing::instrument()]
+pub fn add_user_headers(headers: &mut HeaderMap, user_headers: HeaderMap) {
+    headers.extend(user_headers.into_iter());
+}
+
+#[tracing::instrument()]
+pub fn add_content_type_headers(headers: &mut HeaderMap, name: &str) {
+    let guess = mime_guess::from_path(name);
+    let content_type = guess.first_or_octet_stream().to_string();
+    headers.insert(CONTENT_TYPE, HeaderValue::from_str(&content_type).unwrap());
+}
+
+#[tracing::instrument()]
+pub fn add_content_disposition_headers(
+    headers: &mut HeaderMap,
+    filename: &str,
+    content_path: &str,
+    should_download: bool,
+) -> String {
+    let mut name = get_filename(content_path);
+    if !filename.is_empty() {
+        name = filename.to_string();
+    }
+    if !name.is_empty() {
+        let disposition = if should_download {
+            DISPOSITION_ATTACHMENT
+        } else {
+            DISPOSITION_INLINE
+        };
+        set_content_disposition_headers(headers, &name, disposition);
+    }
+    name
+}
+
+#[tracing::instrument()]
+pub fn set_content_disposition_headers(headers: &mut HeaderMap, filename: &str, disposition: &str) {
+    headers.insert(
+        CONTENT_DISPOSITION,
+        HeaderValue::from_str(&format!("{}; filename={}", disposition, filename)).unwrap(),
+    );
+}
+
+#[tracing::instrument()]
+pub fn add_cache_control_headers(headers: &mut HeaderMap, content_path: String) {
+    if true {
+        // todo(arqu): work out if cpath is mutable
+        // now we just treat everything as mutable
+        // should also utilize the cache flag on config
+        headers.insert(
+            LAST_MODIFIED,
+            HeaderValue::from_str(&Utc::now().to_string()).unwrap(),
+        );
+    } else {
+        headers.insert(LAST_MODIFIED, HeaderValue::from_str("0").unwrap());
+        headers.insert(CACHE_CONTROL, VAL_IMMUTABLE_MAX_AGE.clone());
+    }
+}
+
+#[tracing::instrument()]
+pub fn set_etag_headers(headers: &mut HeaderMap, etag: String) {
```

Comment:
maybe look into https://github.com/magiclen/entity-tag instead of manually impl?

Reply:
So, we're working in a content addressed world so most etags are already baked for us. The only exception is UnixFS dirs, and we should hand roll that anyway. We really don't need any of the features provided here. Only makes our code delegated elsewhere. Your call.